### PR TITLE
ability to generate multiple variations from interactive parameter examples

### PR DESCRIPTION
**Pull request by jtoy**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Does your submission pass tests?

> Note: please do NOT include build files (those generate by the build-xxx commands) with your PR,

Thank you for your help in advance, much appreciated !

### CODE REVIEW

**Comment by kaosat-dev** [comment]
Thanks for this @jtoy  ! I will take it for a spin today.
Not quite sure if it should be a separate file though, but I will try out.

**Comment by kaosat-dev** [comment]
@jtoy taking a closer look at the code: the script creates a different stl output for every possibly parameter combo right ?

**Comment by jtoy** [comment]
yes, that is how I am able to generate variations versions. I am using this code in my production system now.

File: src/cli/multiopenjscad.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env node
+
+#a simple script that reads an interactive params example and generates every variation
+#to use this:  node multiopenjscad.js /path/to/gears.jscad
+#you must have the packages mkdirp and eval installed to use this.
+const fs = require('fs')
+const path = require('path')
+const {execSync} = require('child_process')
+var _eval = require('eval')
```
Comment (kaosat-dev) [inline src/cli/multiopenjscad.js]: why import and then not use ?

File: src/cli/multiopenjscad.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+if(args.length < 1){
+  console.log("pass file to parse")
+  process.exit()
+}
+
+inputFile = args[0]
+ftype = inputFile.split("/").slice(-1)[0].split(".")[0]
+console.log("class: "+ftype)
+let src = fs.readFileSync(inputFile, 'UTF8')
+src += "\ngetParameterDefinitions()\n"
```
Comment (kaosat-dev) [inline src/cli/multiopenjscad.js]: prefer es6 templates

File: src/cli/multiopenjscad.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    array.push(i)
+  }
+  if(array.length > 0){
+    return array
+  }else {
+    return [json['initial']]
+  }
+}
+
+getGetParameterDefinitions = new Function(src + ";\nreturn getParameterDefinitions();");
```
Comment (kaosat-dev) [inline src/cli/multiopenjscad.js]: there are already tools in jscad for extracting parameter definitions for CLI : any reason not to use them?

File: src/cli/multiopenjscad.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+console.log("total variations: "+combos.length)
+for(var i=0;  i < combos.length; i++){
+  params = ""
+  folder = ""
+  // TODO sort params so that the sha is always correct
+  Object.keys(combos[i]).forEach(function(key) {
+    params += ` --${key} ${combos[i][key]} `
+    folder += `${key}_${combos[i][key]}__`
+  })
+  outputPath = `objects/${ftype}/${folder}/`
```
Comment (kaosat-dev) [inline src/cli/multiopenjscad.js]: prefer path creations with require('path').join() as it is cross platform

File: src/cli/multiopenjscad.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env node
+
+#a simple script that reads an interactive params example and generates every variation
+#to use this:  node multiopenjscad.js /path/to/gears.jscad
+#you must have the packages mkdirp and eval installed to use this.
+const fs = require('fs')
+const path = require('path')
+const {execSync} = require('child_process')
+var _eval = require('eval')
+var mkdirp = require('mkdirp');
```
Comment (kaosat-dev) [inline src/cli/multiopenjscad.js]: add a dependency, can it be avoided ?

File: src/cli/multiopenjscad.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+src += "\ngetParameterDefinitions()\n"
+eval(src)
+function to_array(json){
+  array = []
+  if(json.hasOwnProperty('step')){
+    step = json['step']
+  }else{
+    step = 1
+  }
+  for(var i = json['min'];i <= json['max'];i+= step){
```
Comment (kaosat-dev) [inline src/cli/multiopenjscad.js]: the script requires min and max values to be specified for all params right ?
in other words : the params must be finite ?

File: src/cli/multiopenjscad.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+}
+combos = Object.keys(json).reduce((acc, key) => {
+  newArray = [];
+
+  json[key].forEach(item => {
+    if (!acc || !acc.length) { // First iteration
+      newArray.push({[key]: item});
+    } else {
+      acc.forEach(obj => {
+        newArray.push({...obj, [key]: item});
```
Comment (kaosat-dev) [inline src/cli/multiopenjscad.js]: please avoid semicolons , as we use [standard](https://standardjs.com/) styling :)

**Comment by kaosat-dev** [comment]
@jtoy  ok, does that mean that all params need to be finite? (ie, min/max)
Also did you not run into any excessive computation time issues?
This script could be very useful for pre-computing geometry for different param sets to have a 'cached' version (tradeoff of speed vs storage).

I added a review of the code, please do not take the remarks badly!

File: src/cli/multiopenjscad.js (hunk 6 above)
Comment (kaosat-dev) [reply src/cli/multiopenjscad.js]: also favor let over var

**Comment by jtoy** [comment]
Yes, the min and max need to be given, probably should write some notes or not run if no min/max given.
no offense taken, I don't write a lot of javascript.

File: src/cli/multiopenjscad.js (hunk 3 above)
Comment (jtoy) [reply src/cli/multiopenjscad.js]: hi sorry for the late reply, im working on this now. I figured this existed, where can I find this piece?

**Comment by z3dev** [comment]
@jtoy no offense, but i would like to close the pull request without merging.

if you still need this functionality the please make the same changes again branch V2.
thanks

**Comment by z3dev** [comment]
V1 pull requests are no longer accepted. Please try again with V2 compatible changes.

**Pull request closed without merging**